asset-registry/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+		type AuthorityOrigin: EnsureOrigin<<Self as frame_system::Config>::Origin>;
+
+		/// A filter ran upon metadata registration that assigns an is and
+		/// potentially modifies the supplied metadata.
+		type OnRegisterAsset: AssetProcessor<Self::AssetId, AssetMetadata<Self::Balance, Self::CustomMetadata>>;
+
+		/// The balance type.
+		type Balance: Parameter
+			+ Member
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ Into<u128>;
+
+		// /// Weight information for extrinsics in this module.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Asset was not found
+		AssetNotFound,
+		/// The version of the `Versioned` value used is not able to be
+		/// interpreted.
+		BadVersion,
+		/// The asset id is invalid.
+		InvalidAssetId,
+		/// Another asset was already register with this location.
+		ConflictingLocation,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		RegisteredAsset {
+			asset_id: T::AssetId,
+			metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+		},
+		UpdatedAsset {
+			asset_id: T::AssetId,
+			metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+		},
+		SetLocation {
+			asset_id: T::AssetId,
+			location: Box<VersionedMultiLocation>,
+		},
+	}
+
+	/// The total issuance of a token type.
+	#[pallet::storage]
+	#[pallet::getter(fn get_metadata)]
+	pub type Metadata<T: Config> =
+		StorageMap<_, Twox64Concat, T::AssetId, AssetMetadata<T::Balance, T::CustomMetadata>, OptionQuery>;
+
+	/// The total issuance of a token type.
+	#[pallet::storage]
+	#[pallet::getter(fn get_asset_id)]
+	pub type MultiLocationLookup<T: Config> = StorageMap<_, Twox64Concat, MultiLocation, T::AssetId, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn last_asset_id)]
+	pub(crate) type LastAssetId<T: Config> = StorageValue<_, T::AssetId, OptionQuery>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		_phantom: PhantomData<T>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self {
+				_phantom: Default::default(),
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {}
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(T::WeightInfo::register_asset())]
+		#[transactional]
+		pub fn register_asset(
+			origin: OriginFor<T>,
+			metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+			asset_id: Option<T::AssetId>,
+		) -> DispatchResult {
+			let _ = T::AuthorityOrigin::ensure_origin(origin)?;
+
+			let (asset_id, metadata) = T::OnRegisterAsset::process_asset(asset_id, metadata)?;
```

Comment:
I think it'll be good if move `process_asset` processing part into `insert_metdata()`. for parachains who want to use their own dispatch call and intergrate orml, they can use `orml_asset_registry::Pallet::insert_metadata()`

Reply:
That's a bit problematic, since `insert_metadata` is also used from within the `update_asset` function. Should we also call that in `update_asset` and `set_location`?